internal/framework/controller/predicate/nginxproxy.go

```diff
@@ -0,0 +1,89 @@
+package predicate
+
+import (
+	"context"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	ngfAPI "github.com/nginxinc/nginx-gateway-fabric/apis/v1alpha1"
+)
+
+// NginxProxyPredicate implements a predicate function for an NginxProxy resource.
+// This predicate will skip events for NginxProxies that aren't linked to the GatewayClass
+// for this controller.
+type NginxProxyPredicate struct {
+	predicate.Funcs
+	Client         client.Client
+	ControllerName string
+}
+
+// Create implements default CreateEvent filter for validating an NginxProxy resource.
+func (npp NginxProxyPredicate) Create(e event.CreateEvent) bool {
+	if e.Object == nil {
+		return false
+	}
+
+	np, ok := e.Object.(*ngfAPI.NginxProxy)
+	if !ok {
+		return false
+	}
+
+	return npp.validateNginxProxy(np)
+}
+
+// Update implements default UpdateEvent filter for validating an NginxProxy.
+func (npp NginxProxyPredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectNew == nil {
+		return false
+	}
+	npNew, ok := e.ObjectNew.(*ngfAPI.NginxProxy)
+	if !ok {
+		return false
+	}
+
+	return npp.validateNginxProxy(npNew)
+}
+
+// Delete implements default DeleteEvent filter for validating an NginxProxy resource.
+func (npp NginxProxyPredicate) Delete(e event.DeleteEvent) bool {
+	if e.Object == nil {
+		return false
+	}
+
+	np, ok := e.Object.(*ngfAPI.NginxProxy)
+	if !ok {
+		return false
+	}
+
+	return npp.validateNginxProxy(np)
+}
+
+// validateNginxProxy ensures that the NginxProxy resource being processed is the same one
+// that is referenced in the GatewayClass for this controller.
+func (npp NginxProxyPredicate) validateNginxProxy(np *ngfAPI.NginxProxy) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	var gcList v1beta1.GatewayClassList
+	if err := npp.Client.List(ctx, &gcList); err != nil {
+		return false
+	}
+
+	for _, gc := range gcList.Items {
+		if string(gc.Spec.ControllerName) == npp.ControllerName && gc.Spec.ParametersRef != nil {
```

Comment:
What if the NginxProxy CRD is created before the gateway class, or it's not referenced in the gateway class until after it's been created? Will we miss it?

Reply:
Yeah actually, we will.

So this predicate should actually go away, and I think when we build the graph, that's when we determine the NginxProxy we care about.